engine/src/main/java/io/zeebe/engine/processing/job/JobBatchActivateProcessor.java

```diff
@@ -219,12 +205,12 @@ private void rejectCommand(
           "Expected to reject an invalid activate job batch command, but it appears to be valid");
     }
 
-    streamWriter.appendRejection(record, rejectionType, rejectionReason);
+    rejectionWriter.appendRejection(record, rejectionType, rejectionReason);
     responseWriter.writeRejectionOnCommand(record, rejectionType, rejectionReason);
+    responseWriter.flush();
```

Comment:
It's not clear to me why this explicit flush is needed now. What am I missing?

Reply:
That is an excellent question and I am glad you brought it up. I meant to talk to you about this, as I am unhappy myself, unsure how to do it right, and generally interested in discussing some details here.

There are two choices to write the response. I could use the `responseWriter` passed in to `processRecord`, or I could use the `writers.response()` which I get from the `writers` object.

Since I use the writers from the `writers` object for writing state, rejections and commands, I also want to use it for the response. More as a matter of consistency. The first thing that is not ideal is that the response writer has a different style of interface than the other three. The other three have a single method with several parameters to write a single record. The response writer in contrast has many methods with one parameter each, that together compose a response.

So using it - as is - is inconvenient. Better to wrap it in a `TypedResponseWriterImpl` which gives me a nicer interface. Personally, I would prefer it if the `writers` object gave me a `TypedResponseWriterImpl` directly. Right now I cannot imagine ever using the `CommandResponseWriter` directly.

Ok, so now I have my four writers, all from the same source object and all with a somewhat similar interface. 

Now we get to your question:
I need to do the flush explicitly, because when I use the response writer from the `writers` object, then nobody else calls flush on it. 
If I would instead use the response writer passed in to `processRecord`, then the caller of that method calls flush. 

At least this is what I think happens. To be honest, I haven't fully understood how it works.

But lines like `io.zeebe.engine.processing.bpmn.BpmnStreamProcessor#processRecord 86` and `io.zeebe.engine.processing.bpmn.behavior.BpmnWorkflowResultSenderBehavior#sendResult() 78` make me believe that the flush is called for me if I use the response writer passed in as argument.

See also this JavaDoc in `TypedRecordProcessor`:
```
  /**
   * @param position the position of the current record to process
   * @param record the record to process
   * @param responseWriter the default side effect that can be used for sending responses. {@link
   *     TypedResponseWriter#flush()} must not be called in this method.
   * @param streamWriter
   * @param sideEffect consumer to replace the default side effect (response writer). Can be used to
   *     implement other types of side effects or composite side effects. If a composite side effect
   *     involving the response writer is used, {@link TypedResponseWriter#flush()} must be called
   *     in the {@link SideEffectProducer} implementation.
   */
  default void processRecord(
      final long position,
      final TypedRecord<T> record,
      final TypedResponseWriter responseWriter,
      final TypedStreamWriter streamWriter,
      final Consumer<SideEffectProducer> sideEffect) {
    processRecord(record, responseWriter, streamWriter, sideEffect);
  }
```

Note also that there are other processors which also call `flush()` explicitly. See e.g. `io.zeebe.engine.processing.message.MessagePublishProcessor#sendCorrelateCommand`
 
So I have a choice between:
a) using all writers from the `writers` object. Then it is nice and consistent, but I need to call `flush()` manually, thus bringing more responsibility into the class. I also need to create a `TypedResonseWriterImpl` myself and to do that the class needs to know where to find out the partition.
b) using most writers from `writers` object, but using the parameter passed in to `processRecords` to write the response. This is a little inconsistent but closer to the intention of what the `processRecords` method was written for. So the caller will be less surprised if I ignore the response writer passed in as argument and instead use another object on which I even call flush myself.

I don't know which choice is better and I think much depends on how we want to evolve these interfaces. Shall we on balance use the writers passed into the `processRecords` method? or should we ignore these completely and only use the `writers`? And if we use the latter, how can we ensure that subtle implied logic (like calling flush at the right time) can also be migrated to the `writers` object?

Not expecting you have any clear cut answers, right away. But I wanted to raise the issue and know your opinion on the matter.

Reply:
Interesting read. 

I have to start by saying that when creating the `Writers` class, I was not familiar with the ResponseWriter and just picked one. I had no explicit reason to pick this particular one. If there is an interface that is better suited to be used by the Writers class, then by all means feel free to directly pass that one to it.

Considering the flushing, it sounds to me there is a third option which is more similar to the current behavior. Just pass the same response writer to the Writers object as is currently passed to the method. Flush should then just be called on it from the outside right?

Reply:
So there is now an additional commit which does exactly what you suggested. Please have a look at let me know whether you like the flush version better or the version that in the last commit.

Reply:
I prefer the version in the last commit. Nice work!